### PR TITLE
Adapting to new stat architecture

### LINKED ISSUE #1684: Compute only compute basic Statistics in Ehcache (not rates, ratio, etc)
Stats in `StandardEhcacheStatistics ` are not correctly computed (see `Eh107CacheStatisticsMXBean `) and only counters have to be logged and sent to TMS. Not latencies, rate or ratio. Rates and Ratio will be computed in TMS based on its own window and counters in its DB.

@gibsong @anthonydahanne  : here are the steps we discussed yesterday:

- [x] Disable latency stats (in tc-platform, ehcache, m&m) => @mathieucarbou (PR #1692)

- [x] expose stats such as Cache:Hits, OnHeapHits, etc. (work started Geoff’s PR : https://github.com/ehcache/ehcache3/pull/1693/files) - think about creating a class defining hitcounts, etc. that would be used by the JSR107 Mbean and the Collector. It could be an __Ehcache Statistic Service__ that would be used by both management and JSR107 and will contains all rules to compute stats, like it is done currently in JSR107 Mbean. @henri-tremblay 

- [ ] compute other stats in TMS (i.e. rates, ratios, across clients, etc) => @anthonydahanne / @GaryWKeim 

- [x] get rid of other older now unused stats (ratios / rates) in stat registry @henri-tremblay 

- [x] get rid of stats registry, scheduling and sampling and history stats. TMS : would do a call to start the collector - we would not choose metrics anymore - the collector will collect all ehcache raw counters (a counter = stat name + value + timestamp, so lighter than the stat history object we currently have, and moreover, timestamp is the same for all; collector will push all these counters to monitoring buffer => @mathieucarbou (https://github.com/Terracotta-OSS/terracotta-platform/pull/259)

### PR DESCRIPTION
**Pull request by mathieucarbou**
Require: https://github.com/Terracotta-OSS/terracotta-platform/pull/259

See #1684 and https://github.com/Terracotta-OSS/terracotta-platform/issues/217 and email sent to you guys

### CODE REVIEW

**Comment by mathieucarbou** [comment]
Note: waiting for release __5.2.0-pre4__ of tc-platform to be available

File: clustered/integration-test/src/test/java/org/ehcache/clustered/management/AbstractClusteringManagementTest.java [hunk 1]
```diff
@@ -194,8 +180,25 @@ public static void beforeClass() throws Exception {
   @AfterClass
   public static void afterClass() throws Exception {
     if (cacheManager != null && cacheManager.getStatus() == Status.AVAILABLE) {
+
+      System.out.println("====> Asking Ehcache to stop sending stats");
```
Comment (henri-tremblay) [inline clustered/integration-test/src/test/java/org/ehcache/clustered/management/AbstractClusteringManagementTest.java]: System.out

File: clustered/integration-test/src/test/java/org/ehcache/clustered/management/ClusteredStatisticsCountTest.java [hunk 2]
```diff
@@ -61,25 +63,15 @@ public void countTest() throws Exception {
       for (ContextualStatistics stat : stats) {
         if (stat.getContext().contains("cacheName") && stat.getContext().get("cacheName").equals("dedicated-cache-1")) {
 
-          Sample<Long>[] samplesCacheHitCount = stat.getStatistic(CounterHistory.class, "Cache:HitCount").getValue();
-          if(samplesCacheHitCount.length > 0) {
-            cacheHitCount = samplesCacheHitCount[samplesCacheHitCount.length - 1].getValue();
+          System.out.println("stats:");
```
Comment (henri-tremblay) [inline clustered/integration-test/src/test/java/org/ehcache/clustered/management/ClusteredStatisticsCountTest.java]: System.out

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java [hunk 3]
```diff
@@ -1239,6 +1245,9 @@ private void onExpiration(K mappedKey, ValueHolder<V> mappedValue, StoreEventSin
     expirationObserver.end(StoreOperationOutcomes.ExpirationOutcome.SUCCESS);
   }
 
+  /**
+   * Note to users of this method: it can return once the close() method is called.
```
Comment (henri-tremblay) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/AbstractOffHeapStore.java]: I don't understand the comment

File: management/src/main/java/org/ehcache/management/cluster/DefaultClusteringManagementService.java [hunk 4]
```diff
@@ -82,7 +82,7 @@ public void start(ServiceProvider<Service> serviceProvider) {
     // get an ordered executor to keep ordering of management call requests
     this.managementCallExecutor = serviceProvider.getService(ExecutionService.class).getOrderedExecutor(
         configuration.getManagementCallExecutorAlias(),
-        new ArrayBlockingQueue<Runnable>(configuration.getManagementCallQueueSize()));
+        new LinkedBlockingQueue<Runnable>(configuration.getManagementCallQueueSize()));
```
Comment (henri-tremblay) [inline management/src/main/java/org/ehcache/management/cluster/DefaultClusteringManagementService.java]: Why changing to LinkedBlockingQueue?

File: management/src/main/java/org/ehcache/management/providers/statistics/StandardEhcacheStatistics.java [hunk 5]
```diff
@@ -15,79 +15,85 @@
...
+    return statisticRegistry.queryStatistics();
   }
 
-  void dispose() {
-    statisticsRegistry.clearRegistrations();
+  @Override
+  public Collection<StatisticDescriptor> getDescriptors() {
+    return statisticRegistry.getDescriptors();
+
+    //TODO: by registering stats into the registry, descriptors are generated
```
Comment (henri-tremblay) [inline management/src/main/java/org/ehcache/management/providers/statistics/StandardEhcacheStatistics.java]: Remove that

File: management/src/main/java/org/ehcache/management/providers/statistics/StandardEhcacheStatistics.java [hunk 6]
```diff
@@ -15,79 +15,85 @@
...
+    statisticRegistry.registerSize("AllocatedByteSize", ValueStatisticDescriptor.descriptor("allocatedMemory", singleton("tier")));
+    statisticRegistry.registerSize("OccupiedByteSize", ValueStatisticDescriptor.descriptor("occupiedMemory", singleton("tier")));
   }
 
-  Statistic<?, ?> queryStatistic(String fullStatisticName, long since) {
-    return statisticsRegistryMetadata.queryStatistic(fullStatisticName, since);
+  public Statistic<?, ?> queryStatistic(String fullStatisticName) {
+    return statisticRegistry.queryStatistic(fullStatisticName);
+
+    //TODO: register these stats above with .registerCounter(...) or .registerSize(...)
```
Comment (henri-tremblay) [inline management/src/main/java/org/ehcache/management/providers/statistics/StandardEhcacheStatistics.java]: Remova that

File: management/src/test/java/org/ehcache/docs/ManagementTest.java [hunk 7]
```diff
@@ -83,33 +84,33 @@ public void usingManagementRegistry() throws Exception {
...
-
-        // hit count is a sampled stat, for example its values could be [0,0,3,4].
-        // In the present case, only the last value is important to us , the cache was eventually hit 4 times
-        if (onHeapStore_Hit_Count.getValue().length > 0) {
-          int mostRecentIndex = onHeapStore_Hit_Count.getValue().length - 1;
-          onHeapHitCount = onHeapStore_Hit_Count.getValue()[mostRecentIndex].getValue();
-        }
-
-      } while (onHeapHitCount != 4L);
+//      Context context = StatsUtil.createContext(managementRegistry); // <5>
```
Comment (henri-tremblay) [inline management/src/test/java/org/ehcache/docs/ManagementTest.java]: Why is it all commented?

File: management/src/main/java/org/ehcache/management/cluster/DefaultClusteringManagementService.java (hunk 4 above)
Comment (mathieucarbou) [inline management/src/main/java/org/ehcache/management/cluster/DefaultClusteringManagementService.java]: no reason... I tried this when I found a bug in ehcache close that is slow(). I'll change back.

**Comment by mathieucarbou** [comment]
@henri-tremblay @ljacomet : I ran a complete `./gradlew clean build` with the new release __5.2.0-pre5__ of tc-platform and build pass :-)
We just need to wait for the new release to be deployed then.

**Comment by akomakom** [comment]
test this please

**Pull request merged**